# 恐慌与格式字符串

> 原文：<https://golangbyexample.com/panic-format-string-go/>

# **概述**

下面是紧急函数的语法

```go
func panic(v interface{})
```

它以空接口作为参数。它没有提供格式化错误字符串的任何原因。然而，有一个解决办法。 **fmt** 包的 **Sprintf** 功能可用于格式化错误信息，然后将其传递给紧急功能。让我们来看一个程序

# **例**

```go
package main

import (
	"fmt"
)

func main() {
	a := []string{"a", "b"}
	checkAndPrint(a, 2)
	fmt.Println("Exiting normally")
}

func checkAndPrint(a []string, index int) {
	if index > (len(a) - 1) {
		errorString := fmt.Sprintf("Out of bounds access for slice. Index passed: %d", index)
		panic(errorString)
	}
	fmt.Println(a[index])
}
```

**输出**

```go
panic: Out of bounds access for slice. Index passed: 2

goroutine 1 [running]:
main.checkAndPrint(0xc00009af58, 0x2, 0x2, 0x2)
        main.go:17 +0x157
main.main()
        main.go:10 +0x81
exit status 2
```

在上面的程序中，我们有一个函数 **checkAndPrint** ，它在参数中传递的索引处检查并打印切片元素。如果传递的索引大于数组的长度，程序就会恐慌。注意我们如何在将错误字符串传递给**检查和打印**函数中的恐慌之前格式化它

```go
errorString := fmt.Sprintf("Out of bounds access for slice. Index passed: %d", index)
```

该程序还输出正确格式的消息

```go
panic: Out of bounds access for slice. Index passed: 2
```